ql/src/java/org/apache/hadoop/hive/ql/exec/GroupByOperator.java

```diff
@@ -990,24 +955,17 @@ private void flushHashTable(boolean complete) throws HiveException {
     // changed in the future
 
     if (complete) {
-      Iterator<Map.Entry<KeyWrapper, AggregationBuffer[]>> iter = hashAggregations
-          .entrySet().iterator();
-      while (iter.hasNext()) {
-        Map.Entry<KeyWrapper, AggregationBuffer[]> m = iter.next();
-        forward(m.getKey().getKeyArray(), m.getValue());
+      for (Map.Entry<KeyWrapper, AggregationBuffer[]> entry : hashAggregations.entrySet()) {
+        forward(entry.getKey().getKeyArray(), entry.getValue());
       }
-      hashAggregations.clear();
       hashAggregations = null;
-      if (LOG.isInfoEnabled()) {
-        LOG.info("Hash Table completed flushed");
-      }
+      LOG.info("Hash Table completed flushed");
       return;
     }
 
     int oldSize = hashAggregations.size();
-    if (LOG.isInfoEnabled()) {
-      LOG.info("Hash Tbl flush: #hash table = " + oldSize);
-    }
+    LOG.info("Hash Tbl flush: #hash table = " + oldSize);
```

Comment:
We were skipping string garbage for this, I think we can switch to the {} format

Reply:
Thanks @t3rmin4t0r for the review!  I'm not sure the value of that.  Are there many instances where organizations are running with something higher than INFO?  If it's expected to run at INFO, better to just concat and print the message than to have to do the actions of parsing the string, looking for the anchor `{}`, replacing it, etc. etc.  I am happy to do it just for consistency, but from performance, better to simply concat and print.

Reply:
The problem is string concatenation does a lot of work without knowing if it is needed or not -- {} format does not!
http://slf4j.org/faq.html#logging_performance

Reply:
Thanks for chiming in @pgaref 

I love that you're passionate about using the parameters.  I usually like to do this, but it's 50-50 on who likes to see it and who not.  My point is that using parameters:

> can significantly boost logging performance for _disabled_ logging statement.

The problem is that it adds overhead when there are _enabled_ logging statements.  For an INFO level logging, it's almost certainly enabled in a production environment so using the anchor ({}) just adds needless overhead.

Think about it, what is faster?

```
public static void main(String[] args) {
   info1("foo: ", "bar");
   info2("foo: {}", "bar");
}

void info1(String s, String param) {
  System.out.println(s.concat(param));
}

void info2(String s, String param) {
   int index = s.lastIndexOf("{");
   if (index < 0) {
     System.out.println(s);
   } else {
      if (s.lastIndexOf{"}") == (index + 1) {
         String firstHalf= s.subString(0, index);
         String secondHalf = s.subString(index + 2) s.length());
        System.out.println(firstHalf + param + secondHalf);
    } else {
        System.out.println(s);
    }
}
```

Clearly the second option is going to be slower.  `Info1` generates a string certainly, and `Info2` does not.  However, if the application is almost always going to be running at the INFO level, it doesn't really make a lot of sense to have to always go down `Info2`.

That said, there is certainly value in having consistency (always using parameters), and I think messages are easier to read when formatted this way than when they are all piecemeal with the "+" signs, but if you want to understand the raw performance, then what is currently there is preferable in an environment where the log level is set to a minimum of INFO (which almost every environment I've worked on is).

Reply:
(see below for more)

Reply:
Hey @belugabehr -- thanks for the detailed comment!
Tbh I was under the impression that the flush size message above was on debug mode -- we are on the same page on this!

There are several places in the project were we check `if (LOG.isDebugEnabled())` before doing the logging, however, this is completely redundant (if we have less than 3 arguments as you pointed out correctly below).
For Log.info mode, as you said, it is more like a matter of consistency.